include/irr/asset/CTexturePacker.h

```diff
@@ -4,35 +4,96 @@
 #include "irr/asset/format/EFormat.h"
 #include "irr/asset/ICPUImage.h"
 #include "irr/core/alloc/GeneralpurposeAddressAllocator.h"
+#include "irr/core/math/morton.h"
 
 namespace irr {
 namespace asset
 {
 
+//TODO move this to different file
+template<typename addr_type>
+class CSquareAddressAllocator
+{
+protected:
+    using lin_addr_alctr_t = core::GeneralpurposeAddressAllocator<addr_type>;
+
+    const uint32_t ADDR_LAYER_SHIFT;
+    const uint32_t ADDR_MORTON_MASK;
+
+    core::smart_refctd_dynamic_array<uint8_t> m_alctrReservedSpace;
+    lin_addr_alctr_t m_addrAlctr;
+
+public:
+    CSquareAllocator(uint32_t _addrBitCntPerDim, uint32_t _squareSz) : 
+        ADDR_LAYER_SHIFT(_addrBitCntPerDim*2u),
+        ADDR_MORTON_MASK((1u<<ADDR_LAYER_SHIFT)-1u),
+        m_alctrReservedSpace(core::make_refctd_dynamic_array<decltype(m_alctrReservedSpace)>(
+            lin_addr_alctr_t::reserved_size(_squareSz*_squareSz, _squareSz*_squareSz, 1ull))
+        ),
+        m_addrAlctr(reinterpret_cast<void*>(m_alctrReservedSpace->data()), 0u, 0u, _squareSz*_squareSz, _squareSz*_squareSz, 1u)
+    {}
+
+    inline addr_type alloc_addr(size_t _subsquareSz)
```

Comment:
if you want a shorthand for allocation make a utility function, but this will no longer be an address allocator if it doesn't conform to the function signatures address_allocator_trais expect.

Reply:
something really inconsistent is going on with your use of `_subsquareSz` across alloc and free

Reply:
no, why?

Reply:
one place uses the `_subsquareSz` as normal and `free` squares it instead?

Also see my comment about this class not being a proper address allocator because it doesn't conform to the API

Reply:
i know it isnt, therefore i dont know why would i want to inherit from general purpose addr allocator instead of having it as member?

Reply:
I'm questioning two things:

1.
> one place uses the _subsquareSz as normal and free squares it instead?

are you sure that is not a bug?


2. Why do you need a class that looks like an address allocator, has methods almost like one and then needlessly wraps an existing address allocator?

Reply:
1. both squares it, maybe u're looking at some old code
2. maybe u're right.. i'll try to rewrite it with inheritance

Reply:
> maybe u're right.. i'll try to rewrite it with inheritance

I think there's probably no need to have it in the first place, since its a very thin intermediatory to `GeneralPurpose`